cpp/src/search/search.cu

```diff
@@ -93,143 +64,161 @@ std::unique_ptr<column> search_ordered(table_view const& t,
-  auto const val_flattened =
-    structs::detail::flatten_nested_columns(scalar_table, {}, {}, flatten_nullability);
-
-  // The struct scalar only contains the struct member columns.
-  // Thus, if there is any null in the input, we must exclude the first column in the flattened
-  // table of the input column from searching because that column is the materialized bitmask of
-  // the input structs column.
-  auto const col_flattened_content  = col_flattened.flattened_columns();
-  auto const col_flattened_children = table_view{std::vector<column_view>{
-    col_flattened_content.begin() + static_cast<size_type>(has_null_elements),
-    col_flattened_content.end()}};
-
-  auto const d_col_children_ptr = table_device_view::create(col_flattened_children, stream);
-  auto const d_val_ptr          = table_device_view::create(val_flattened, stream);
-
-  auto const start_iter = thrust::make_counting_iterator<size_type>(0);
-  auto const end_iter   = start_iter + col.size();
-  auto const comp       = row_equality_comparator(
-    nullate::DYNAMIC{has_null_elements}, *d_col_children_ptr, *d_val_ptr, null_equality::EQUAL);
-  auto const found_iter = thrust::find_if(
-    rmm::exec_policy(stream), start_iter, end_iter, [comp] __device__(auto const idx) {
-      return comp(idx, 0);  // compare col[idx] == val[0].
-    });
-
-  return found_iter != end_iter;
-}
+  template <typename Type>
+  std::enable_if_t<is_nested<Type>(), bool> operator()(column_view const& col,
+                                                       scalar const& value,
+                                                       rmm::cuda_stream_view stream)
+  {
+    CUDF_EXPECTS(col.type() == value.type(), "scalar and column types must match");
+
+    auto constexpr is_struct_type = std::is_same_v<Type, cudf::struct_view>;
+    if constexpr (is_struct_type) {  // struct type ================================================
+      auto const scalar_tview = dynamic_cast<struct_scalar const*>(&value)->view();
+      CUDF_EXPECTS(col.num_children() == scalar_tview.num_columns(),
+                   "struct scalar and structs column must have the same number of children");
+      for (size_type i = 0; i < col.num_children(); ++i) {
+        CUDF_EXPECTS(col.child(i).type() == scalar_tview.column(i).type(),
+                     "scalar and column children types must match");
+      }
+
+      // Generate a column_view of one row having children given from the input scalar.
+      auto const val_col =
+        column_view(data_type{type_id::STRUCT},
+                    1,
+                    nullptr,
+                    nullptr,
+                    0,
+                    0,
+                    std::vector<column_view>{scalar_tview.begin(), scalar_tview.end()});
+
+      return check_contain_for_nested_type(col, val_col, stream);
+    } else {  // list type =========================================================================
```

Comment:
This is precisely the kind of code the type_dispatcher was create to replace. These are entirely distinct code paths and should be in separate dispatched overloads.